clients/src/main/java/org/apache/kafka/common/internals/SecurityManagerCompatibility.java

```diff
@@ -0,0 +1,355 @@
+        }
+
+        @Override
+        public <T> T callAs(Subject subject, Callable<T> action) throws CompletionException {
+            throw createException("Unable to find suitable Subject#doAs or Subject#callAs implementation");
+        }
+    }
+
+    /**
+     * This strategy combines the functionality of the {@link Legacy}, {@link Modern}, and {@link Unsupported}
+     * strategies to provide the legacy APIs as long as they are present and not degraded. If the legacy APIs are
+     * missing or degraded, this falls back to the modern APIs.
+     */
+    class Composite implements SecurityManagerCompatibility {
+
+        private static final Logger log = LoggerFactory.getLogger(Composite.class);
+        private static final Composite INSTANCE = new Composite(Loader.forName());
+
+        private final SecurityManagerCompatibility fallbackStrategy;
+        private final AtomicReference<SecurityManagerCompatibility> activeStrategy;
+
+        // Visible for testing
+        Composite(Loader loader) {
+            SecurityManagerCompatibility initial;
+            SecurityManagerCompatibility fallback = null;
+            try {
+                initial = new Legacy(loader);
+                try {
+                    fallback = new Modern(loader);
+                    // This is expected for JRE 18+
+                    log.debug("Loaded legacy SecurityManager methods, will fall back to modern methods after UnsupportedOperationException");
+                } catch (NoSuchMethodException | ClassNotFoundException ex) {
+                    // This is expected for JRE <= 17
+                    log.debug("Unable to load modern Subject methods, relying only on legacy methods", ex);
+                }
+            } catch (ClassNotFoundException | NoSuchMethodException e) {
+                try {
+                    initial = new Modern(loader);
+                    // This is expected for JREs after the removal takes place.
+                    log.debug("Unable to load legacy SecurityManager methods, relying only on modern methods", e);
+                } catch (NoSuchMethodException | ClassNotFoundException ex) {
+                    initial = new Unsupported(e, ex);
+                    // This is not expected in normal use, only in test environments.
+                    log.error("Unable to load legacy SecurityManager methods", e);
+                    log.error("Unable to load modern Subject methods", ex);
+                }
+            }
+            Objects.requireNonNull(initial, "initial strategy must be defined");
+            activeStrategy = new AtomicReference<>(initial);
+            fallbackStrategy = fallback;
+        }
+
+        private <T> T performAction(Function<SecurityManagerCompatibility, T> action) {
+            SecurityManagerCompatibility active = activeStrategy.get();
+            try {
+                return action.apply(active);
+            } catch (UnsupportedOperationException e) {
```

Comment:
Is it possible that an `UnsupportedOperationException` gets thrown directly by the underlying action, and not because we intercepted an `IllegalAccessException`?

Reply:
Yes, and this is intentional. JDK23 has started to throw UnsupportedOperationException from these methods to degrade them, rather than just remove them entirely.